zetta_utils/augmentations/misalign.py

```diff
@@ -0,0 +1,162 @@
+    The selected displacement will be rounded to always be divisible
+    by `disp_in_unit_must_be_divisible_by`, which should correspond to
+    the lowest resolution of the layers being cropped.
+
+    Data handling is implemented through recording the fraction
+    of the misalignment relative to the total bbox size.
+    """
+
+    prob: float
+    disp_min_in_unit: float
+    disp_max_in_unit: float
+    disp_in_unit_must_be_divisible_by: float | None = None
+    mode: Literal["slip", "step"] = "slip"
+    keys_to_apply: list[str] | None = None
+
+    prepared_disp_fraction: tuple[float, float] | None = attrs.field(init=False, default=None)
+
+    def process_index(
+        self, idx: VolumetricIndex, mode: Literal["read", "write"]
+    ) -> VolumetricIndex:
+        if mode != "read":
+            raise NotImplementedError()
+        disp_x_in_unit_magn = random.uniform(self.disp_min_in_unit, self.disp_max_in_unit)
+        disp_y_in_unit_magn = random.uniform(self.disp_min_in_unit, self.disp_max_in_unit)
+        if self.disp_in_unit_must_be_divisible_by is not None:
+            disp_x_in_unit_magn = math.floor(
+                disp_x_in_unit_magn / self.disp_in_unit_must_be_divisible_by
+            )
+            disp_y_in_unit_magn = math.floor(
+                disp_y_in_unit_magn / self.disp_in_unit_must_be_divisible_by
+            )
+        disp_x_in_unit = disp_x_in_unit_magn * random.choice([1, -1])
+        disp_y_in_unit = disp_y_in_unit_magn * random.choice([1, -1])
+
+        disp_x_in_idx_res = disp_x_in_unit / idx.resolution[0]
+        disp_y_in_idx_res = disp_y_in_unit / idx.resolution[1]
+
+        # self.prepared_disp = (disp_x, disp_y)
+
+        start_offset = [0, 0, 0]
+        end_offset = [0, 0, 0]
+        if disp_x_in_idx_res > 0:
+            start_offset[0] = -disp_x_in_idx_res
+        else:
+            end_offset[0] = -disp_x_in_idx_res
+        if disp_y_in_idx_res > 0:
+            start_offset[1] = -disp_y_in_idx_res
+        else:
+            end_offset[1] = -disp_y_in_idx_res
+
+        idx = idx.translated_start(Vec3D[int](*start_offset)).translated_end(
+            Vec3D[int](*end_offset)
+        )
+        self.prepared_disp_fraction = (
+            disp_x_in_idx_res / idx.shape[0],
+            disp_y_in_idx_res / idx.shape[0],
+        )
+        return idx
+
+    def process_data(self, data: T, mode: Literal["read", "write"]) -> T:
+        if mode != "read":
+            raise NotImplementedError()
+
+        if isinstance(data, torch.Tensor):
+            z_size = data.shape[-1]
+        else:
+            if not self.keys_to_apply:
+                raise ValueError(
+                    "`keys_to_apply` must be a non-empty list of strings when "
+                    "applying to data of type `dict`"
+                )
+            sizes = [data[k].shape for k in self.keys_to_apply]
+            if not all(e[-1] == sizes[0][-1] for e in sizes):
```

Comment:
I typically use an output patch size smaller than the input patch size, so we should allow varying z sizes.